src/server/game/Entities/Unit/Unit.cpp

```diff
@@ -12402,65 +12368,42 @@ void Unit::SetStunned(bool apply)
             RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_STUNNED);
 
         if (!HasUnitState(UNIT_STATE_ROOT))         // prevent moving if it also has root effect
-        {
-            WorldPacket data(SMSG_FORCE_MOVE_UNROOT, 8+4);
-            data << GetPackGUID();
-            data << uint32(0);
-            SendMessageToSet(&data, true);
-
-            RemoveUnitMovementFlag(MOVEMENTFLAG_ROOT);
-        }
+            SetRooted(false);
     }
 }
 
 void Unit::SetRooted(bool apply)
 {
-    if (apply)
-    {
-        if (m_rootTimes > 0) // blizzard internal check?
-            m_rootTimes++;
+    // do nothing if the unit is already in the required state
+    if ((HasUnitMovementFlag(MOVEMENTFLAG_ROOT) || HasUnitMovementFlag(MOVEMENTFLAG_PENDING_ROOT)) == apply && !HasPendingMovementChange(ROOT))
+        return;
 
-        // MOVEMENTFLAG_ROOT cannot be used in conjunction with MOVEMENTFLAG_MASK_MOVING (tested 3.3.5a)
-        // this will freeze clients. That's why we remove MOVEMENTFLAG_MASK_MOVING before
-        // setting MOVEMENTFLAG_ROOT
-        RemoveUnitMovementFlag(MOVEMENTFLAG_MASK_MOVING);
-        AddUnitMovementFlag(MOVEMENTFLAG_ROOT);
-        StopMoving();
+    if (IsMovedByPlayer() && IsInWorld())
+        MovementPacketSender::SendMovementFlagChangeToMover(this, MOVEMENTFLAG_ROOT, apply);
+    else if (IsMovedByPlayer() && !IsInWorld())
+        SetRootedReal(apply);
+    else
+    {
+        if(apply)
+            StopMoving(); // @todo: this method needs a rework. until then, should it be called only for NPC or also for player-controlled units?
```

Comment:
Should be called also for players.
If don't call it for player - https://www.youtube.com/watch?v=uCV_4yVNVfk&feature=youtu.be